mllib/src/main/scala/org/apache/spark/mllib/fpm/FPGrowth.scala

```diff
@@ -0,0 +1,208 @@
+ * independent group of mining tasks. More detail of this algorithm can be found at
+ * http://infolab.stanford.edu/~echang/recsys08-69.pdf
+ */
+class FPGrowth private(private var minSupport: Double) extends Logging with Serializable {
+
+  /**
+   * Constructs a FPGrowth instance with default parameters:
+   * {minSupport: 0.5}
+   */
+  def this() = this(0.5)
+
+  /**
+   * set the minimal support level, default is 0.5
+   * @param minSupport minimal support level
+   */
+  def setMinSupport(minSupport: Double): this.type = {
+    this.minSupport = minSupport
+    this
+  }
+
+  /**
+   * Compute a FPGrowth Model that contains frequent pattern result.
+   * @param data input data set
+   * @return FPGrowth Model
+   */
+  def run(data: RDD[Array[String]]): FPGrowthModel = {
+    val model = runAlgorithm(data)
+    model
+  }
+
+  /**
+   * Implementation of PFP.
+   */
+  private def runAlgorithm(data: RDD[Array[String]]): FPGrowthModel = {
+    val count = data.count()
+    val minCount = minSupport * count
+    val single = generateSingleItem(data, minCount)
+    val combinations = generateCombinations(data, minCount, single)
+    new FPGrowthModel(single ++ combinations)
+  }
+
+  /**
+   * Generate single item pattern by filtering the input data using minimal support level
+   */
+  private def generateSingleItem(
+      data: RDD[Array[String]],
+      minCount: Double): Array[(String, Int)] = {
+    data.flatMap(v => v)
+      .map(v => (v, 1))
+      .reduceByKey(_ + _)
+      .filter(_._2 >= minCount)
+      .collect()
+      .distinct
+      .sortWith(_._2 > _._2)
+  }
+
+  /**
+   * Generate combination of items by computing on FPTree,
+   * the computation is done on each FPTree partitions.
+   */
+  private def generateCombinations(
+      data: RDD[Array[String]],
+      minCount: Double,
+      singleItem: Array[(String, Int)]): Array[(String, Int)] = {
+    val single = data.context.broadcast(singleItem)
+    data.flatMap(basket => createFPTree(basket, single))
+      .groupByKey()
+      .flatMap(partition => runFPTree(partition, minCount))
+      .collect()
+  }
+
+  /**
+   * Create FP-Tree partition for the giving basket
+   */
+  private def createFPTree(
+      basket: Array[String],
+      singleItem: Broadcast[Array[(String, Int)]]): Array[(String, Array[String])] = {
+    var output = ArrayBuffer[(String, Array[String])]()
+    var combination = ArrayBuffer[String]()
+    val single = singleItem.value
+    var items = ArrayBuffer[(String, Int)]()
+
+    // Filter the basket by single item pattern
+    val iterator = basket.iterator
+    while (iterator.hasNext){
+      val item = iterator.next
+      val opt = single.find(_._1.equals(item))
+      if (opt != None) {
+        items ++= opt
+      }
+    }
+
+    // Sort it and create the item combinations
+    val sortedItems = items.sortWith(_._1 > _._1).sortWith(_._2 > _._2).toArray
```

Comment:
Why sorting twice? The second will overwrite the first. Besides, using `sortBy(-_._2)` would be better.